sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -84,91 +79,189 @@ private[engine] final class Preprocessor(
-      .toList
+    }
 
-  private[this] def pullPackages(
-      pkgIds: List[(Ref.PackageId, language.Reference)]
-  ): Result[Unit] =
-    pkgIds match {
-      case (pkgId, context) :: rest =>
-        ResultNeedPackage(
-          pkgId,
-          {
-            case Some(pkg) =>
-              compiledPackages.addPackage(pkgId, pkg).flatMap(_ => pullPackages(rest))
-            case None =>
-              ResultError(Error.Package.MissingPackage(pkgId, context))
-          },
-        )
-      case Nil =>
-        ResultDone.Unit
+  private[this] def collectPackagesInCmd(cmd: ReplayCommand): Set[language.Reference] =
+    cmd match {
+      case ReplayCommand.Create(tmplId, _) =>
+        Set(language.Reference.Template(tmplId))
+      case ReplayCommand.Exercise(tmplId, mbIfaceId, _, _, _) =>
+        mbIfaceId match {
+          case Some(ifaceId) =>
+            Set(language.Reference.InterfaceInstance(tmplId, ifaceId))
+          case None =>
+            Set(language.Reference.Template(tmplId))
+        }
+      case ReplayCommand.ExerciseByKey(tmplId, _, choiceId, _) =>
+        Set(language.Reference.TemplateChoice(tmplId, choiceId))
+      case ReplayCommand.Fetch(tmplId, mbIfaceId, _) =>
+        mbIfaceId match {
+          case Some(ifaceId) =>
+            Set(language.Reference.InterfaceInstance(tmplId, ifaceId))
+          case None =>
+            Set(language.Reference.Template(tmplId))
+        }
+      case ReplayCommand.FetchByKey(tmplId, _) =>
+        Set(language.Reference.TemplateKey(tmplId))
+      case ReplayCommand.LookupByKey(tmplId, _) =>
+        Set(language.Reference.TemplateKey(tmplId))
     }
 
-  private[this] def pullTypePackages(typ: Ast.Type): Result[Unit] =
-    collectNewPackagesFromTypes(List(typ)).flatMap(pullPackages)
+  private[this] def collectPackagesInNodes(nodes: Iterable[Node]): Set[language.Reference] =
+    nodes.collect {
+      case create: Node.Create =>
+        language.Reference.Template(create.templateId)
+      case fetch: Node.Fetch =>
+        fetch.interfaceId match {
+          case Some(ifaceId) =>
+            language.Reference.InterfaceInstance(fetch.templateId, ifaceId)
+          case None =>
+            language.Reference.Template(fetch.templateId)
+        }
+      case lookup: Node.LookupByKey =>
+        language.Reference.TemplateKey(lookup.templateId)
+      case exe: Node.Exercise =>
+        exe.interfaceId match {
+          case Some(ifaceId) =>
+            language.Reference.InterfaceInstance(exe.templateId, ifaceId)
+          case None =>
+            language.Reference.Template(exe.templateId)
+        }
+    }.toSet
+
+  private[this] def collectPkgInDisclosure(
+      contracts: ImmArray[command.DisclosedContract]
+  ): Set[language.Reference] =
+    contracts.iterator
+      .map(contract => language.Reference.Template(TypeConRef.fromIdentifier(contract.templateId)))
+      .toSet
 
-  private[this] def pullPackage(
+  private[this] def pullPackages(
```

Comment:
Took me a while to realize that compiledPackages was mutable and this was having a side effect. Maybe add a small comment along the lines of "populates compiledPackages".